CIP-0022/CIP-0022.md

```diff
@@ -0,0 +1,113 @@
+---
+CIP: 22
+Title: Pool operator verification
+Authors: Andrew Westberg <andrewwestberg@gmail.com>, Martin Lang <martin@martinlang.at>, Ola Ahlman <ola@ahlnet.nu>
+Comments-URI: no comments yet
+Status: Draft
+Type: Informational
+Created: 2021-06-21
+License: CC-BY-4.0
+---
+
+# Abstract
+
+This proposal describes a method allowing a stakepool operator to provide credentials to verify that they are the rightful manager for their stakepool.
+
+# Motivation
+
+Many websites such pooltool.io, adapools.org, and others need to allow pool operators special access to modify the way their pool appears on the website. SPOCRA and other organizations also have a need to allow voting on proposals and ensure that each vote cast is from a valid pool operator. Today, these sites and organizations all use different techniques for validating pool operators.
+
+pooltool.io - Validates operators by receiving 1 ada spent from the pool's registered rewards account
+
+adapools.org - Validates operators by requesting that the operator include a special generated value in their extended pool metadata json file.
+
+This proposal desires to simplify and streamline a single approach that all can reference in order to verify that a pool operator is who they say they are.
+
+Current Tools that have implemented it already or have plans to implement it:
+* [CNCLI](https://github.com/AndrewWestberg/cncli)
+* [JorManager](https://bitbucket.org/muamw10/jormanager/)
+* [StakePoolOperator Scripts](https://github.com/gitmachtl/scripts)
+* [CNTools](https://cardano-community.github.io/guild-operators/#/Scripts/cntools)
+
+
+# Specification
+
+In order to achieve the goals of this CIP, the pool operator needs to provide some credential or credentials to the validating party which cannot be spoofed. The VRF pool keys and VRF signature algorithm implemented in libsodium are chosen to build and provide this credential/signature. This signature can then be validated and the operator verified without ever exposing any of the pool's private key information. This technique is very similar to verifying that a block produced by another pool is valid. The only difference is that instead of validating the slot seed for a given pool, we're validating a pre-determined message hash.
```

Comment:
I think it would be best to structure the specification in two parts:

- Some clear steps 
- Code Examples

From the code examples, the steps seems to roughly be:

1. Validating server sends a challenge to the pool operator
2. The pool operator hashes the challenge (blake2b, digest-length=??) + surrounding structure (see other comment)
3. The pool operator produces a signature of the challenge digest using its VRF signing key 
4. The pool operator forwards the signature to the validating server (it is not needed to send the VRF public key since the validating server is already validating against some VRF key identifying the operator)
5. The validating server verifies the signature by following step 2. and using the VRF key of the operator.

Reply:
> (it is not needed to send the VRF public key since the validating server is already validating against some VRF key identifying the operator)

How does the Validating server get the VRF vkey? The `cardano-cli query pool-params` only gives us the hash of the VRF vkey. We can get the VRF vkey from minted blocks, but that only works to validate a pool that has minted at least one block. Is there a way to retrieve the VRF vkey from the chain for a pool that has not yet minted a block?

Reply:
@KtorZ I've re-structured to add a Verification Steps section before the code examples.

I've also included the step of the pool operator sending the VRF public key to the validating server. In most cases, this will not be required as the Validating Server can just get it from a minted block on the blockchain. However, in the edge case of trying to validate a new pool that has yet to mint a block, it will be required. The Validating server can then check this vrf.vkey against the hashed vrf.vkey that exists in the pool's regcert on chain.

Reply:
Err. Good point, I somehow expected the validating server to already have a hold on the VRF key it's trying to validating. Though in practice, the verification request will likely be initiated by a pool operator claiming to be the owner of some key.